kombu/utils/objects.py

```diff
@@ -1,64 +1,54 @@
 """Object Utilities."""
 
+__all__ = ('cached_property',)
 
-class cached_property:
-    """Cached property descriptor.
+try:
+    from functools import _NOT_FOUND, cached_property as _cached_property
+except ImportError:
+    # TODO: Remove this fallback once we drop support for Python < 3.8
+    from cached_property import threaded_cached_property as _cached_property
 
-    Caches the return value of the get method on first call.
+    _NOT_FOUND = object()
 
-    Examples:
-        .. code-block:: python
 
-            @cached_property
-            def connection(self):
-                return Connection()
-
-            @connection.setter  # Prepares stored value
-            def connection(self, value):
-                if value is None:
-                    raise TypeError('Connection must be a connection')
-                return value
-
-            @connection.deleter
-            def connection(self, value):
-                # Additional action to do at del(self.attr)
-                if value is not None:
-                    print('Connection {0!r} deleted'.format(value)
-    """
-
-    def __init__(self, fget=None, fset=None, fdel=None, doc=None):
-        self.__get = fget
+class cached_property(_cached_property):
+    def __init__(self, fget=None, fset=None, fdel=None):
+        super().__init__(fget)
         self.__set = fset
         self.__del = fdel
-        self.__doc__ = doc or fget.__doc__
-        self.__name__ = fget.__name__
-        self.__module__ = fget.__module__
 
-    def __get__(self, obj, type=None):
-        if obj is None:
-            return self
-        try:
-            return obj.__dict__[self.__name__]
-        except KeyError:
-            value = obj.__dict__[self.__name__] = self.__get(obj)
-            return value
+        if not hasattr(self, 'attrname'):
+            # This is a backport so we set this ourselves.
+            self.attrname = self.func.__name__
+
+    def __get__(self, instance, owner=None):
+        # TODO: Remove this after we drop support for Python<3.8
+        #  or fix the signature in the cached_property package
+        return super().__get__(instance, owner)
 
-    def __set__(self, obj, value):
-        if obj is None:
+    def __set__(self, instance, value):
+        if instance is None:
             return self
-        if self.__set is not None:
-            value = self.__set(obj, value)
-        obj.__dict__[self.__name__] = value
 
-    def __delete__(self, obj, _sentinel=object()):
-        if obj is None:
+        with self.lock:
+            if self.__set is not None:
+                value = self.__set(instance, value)
+
+            cache = instance.__dict__
+            cache[self.attrname] = value
```

Comment:
**[question]**: Why do we have to override `set`? `functools.cached_property` and `cached_property` both implement only `get` to calculate the value only once.

Do we really need to re-define the value on cached  properties?

Reply:
Because that's how it's implemented and used in Celery.

Reply:
@xirdneh Any thoughts?